File: App/src/Adaptive.ReactiveTrader.Client.Domain/Repositories/ReferenceDataRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Domain.Models;
using Adaptive.ReactiveTrader.Client.Domain.ServiceClients;

namespace Adaptive.ReactiveTrader.Client.Domain.Repositories
{
    class ReferenceDataRepository : IReferenceDataRepository
    {
        private readonly IReferenceDataServiceClient _referenceDataServiceClient;
        private readonly ICurrencyPairUpdateFactory _currencyPairUpdateFactory;

        public ReferenceDataRepository(IReferenceDataServiceClient referenceDataServiceClient, ICurrencyPairUpdateFactory currencyPairUpdateFactory)
        {
            _referenceDataServiceClient = referenceDataServiceClient;
            _currencyPairUpdateFactory = currencyPairUpdateFactory;
        }

        public IObservable<IEnumerable<ICurrencyPairUpdate>> GetCurrencyPairs()
        {
            return _referenceDataServiceClient.GetCurrencyPairUpdates()
                //.Select(updates => updates.Where(update => update.UpdateType == UpdateTypeDto.Added))
                .Where(updates => updates.Any())
                .Select(updates => updates.Select(update => _currencyPairUpdateFactory.Create(update)))
                .Publish()
                .RefCount();
            }
    }
}

File: App/src/Adaptive.ReactiveTrader.Client.JS/Repositories/ReferenceDataRepository.ts
class ReferenceDataRepository implements IReferenceDataRepository {
    private _referenceDataServiceClient: IReferenceDataServiceClient;
    private _currencyPairUpdateFactory: ICurrencyPairUpdateFactory;

    constructor(
        referenceDataServiceClient: IReferenceDataServiceClient,
        currencyPairUpdateFactory: ICurrencyPairUpdateFactory) {

        this._currencyPairUpdateFactory = currencyPairUpdateFactory;
        this._referenceDataServiceClient = referenceDataServiceClient;
    }
    
    getCurrencyPairs(): Rx.Observable<ICurrencyPairUpdate[]> {
        return Rx.Observable.defer(()=> this._referenceDataServiceClient.getCurrencyPairUpdates())
            .where(updates=> updates.length > 0)
            .select(updates=> this.createCurrencyPairUpdates(updates))
            .catch(()=> Rx.Observable.return([]))
            .repeat()
            .publish()
            .refCount();
    }

    private createCurrencyPairUpdates(updates: CurrencyPairUpdateDto[]) : ICurrencyPairUpdate[] {
        var result = [];

        for (var i = 0; i < updates.length; i++) {
            result[i] = this._currencyPairUpdateFactory.create(updates[i]);
        }

        return result;
    }
} 

